gips/data/sentinel2/sentinel2.py

```diff
@@ -335,15 +431,98 @@ def read_raw(self):
         return image
 
 
-    def _time_report(self, msg, reset_clock=False):
+    def _time_report(self, msg, reset_clock=False, verbosity=None):
         """Provide the user with progress reports, including elapsed time.
 
-        Reset elapsed time with reset_clock=True.
+        Reset elapsed time with reset_clock=True; when starting or
+        resetting the clock, specify a verbosity, or else accept the
+        default of 3.
         """
         start = getattr(self, '_time_report_start', None)
         if reset_clock or start is None:
             start = self._time_report_start = datetime.datetime.now()
-        utils.verbose_out('{}:  {}'.format(datetime.datetime.now() - start, msg, 3))
+            self._time_report_verbosity = 3 if verbosity is None else verbosity
+        elif verbosity is not None:
+            raise ValueError('Changing verbosity is only permitted when resetting the clock')
+        utils.verbose_out('{}:  {}'.format(
+            datetime.datetime.now() - start, msg, self._time_report_verbosity))
+
+
+    def process_rad(self, asset_type, proto_product, sensor, product, filename):
+        """Reverse-engineer TOA ref data back into a TOA radiance product."""
+        self._time_report('Starting TOA radiance processing')
+        asset_instance = self.assets[asset_type] # sentinel2Asset
+        colors = asset_instance._sensors[sensor]['colors']
+
+        radiance_factors = asset_instance.radiance_factors()
+
+        rad_image = gippy.GeoImage(proto_product)
+
+        for i in range(len(rad_image)):
+            color = rad_image[i].Description()
+            rf = radiance_factors[colors.index(color)]
+            self._time_report(
+                'TOA radiance conversion factor for {} (band {}): {}'.format(color, i + 1, rf))
+            rad_image[i] *= rf
```

Comment:
Looks like `GeoRaster operator*=(double val)` doesn't exist and swig doesn't barf on its absence.